Add Java Platform Module System (JPMS) support

cowwoc opened the pull request:
## Summary
This PR adds Java Platform Module System (JPMS) support to Truth while maintaining full backward compatibility with Java 8.

## Changes
- **Core module**: Added `module-info.java` with exports for `com.google.common.truth`
- **Extension modules**: Added module descriptors for:
  - `com.google.truth.extensions.proto` 
  - `com.google.truth.extensions.liteproto`
  - `com.google.truth.extensions.re2j`
- **Multi-release JAR**: Configured Maven compiler and JAR plugins for multi-release compilation
- **Dependencies**: Properly declared module dependencies including Guava, JUnit, JSpecify, Error Prone, and ASM

## Benefits
- JPMS compatibility for modern Java applications (Java 9+)
- Enhanced module encapsulation and security
- Clear dependency boundaries
- Maintains Java 8 compatibility through multi-release JAR structure
- Supports both modular and non-modular usage patterns

## Implementation Details
- Uses multi-release JAR approach with `module-info.java` in `src/main/java9/`
- Static dependencies for optional libraries (ASM, JSpecify, Error Prone)
- Automatic module names for non-modular dependencies
- Maven compiler plugin configured for Java 9+ module compilation

## Test plan
- [x] Core module compiles successfully with Java 9+ module system
- [x] Multi-release JAR structure verified (`META-INF/versions/9/module-info.class`)
- [x] Module dependencies resolve correctly (`--describe-module` verification)
- [x] Java 8 compatibility maintained through multi-release structure
- [x] Line ending normalization applied

cowwoc committed:
Add Java Platform Module System (JPMS) support

- Add module-info.java descriptors for core and extension modules
- Configure multi-release JAR compilation for Java 9+ compatibility
- Maintain Java 8 backward compatibility through multi-release structure
- Support for Guava, JUnit, JSpecify, Error Prone, and ASM dependencies
- Extension modules: truth-proto-extension, truth-liteproto-extension, truth-re2j-extension

cpovirk commented on core/pom.xml:
```diff
@@ -98,6 +98,23 @@
+              </compileSourceRoots>
+              <multiReleaseOutput>true</multiReleaseOutput>
+              <useIncrementalCompilation>false</useIncrementalCompilation>
+              <moduleVersion>1.0</moduleVersion>
```
Would it make sense to use `${project.version}` here? (I remain relatively unfamiliar with modules.) It looks like we don't set a version at all [for Guava](https://github.com/google/guava/blob/b94bd32f1923996261c0363b0a57283be391d87d/guava/pom.xml#L112), but presumably that just means that we should fix Guava, too :)

cowwoc replied:
@cpovirk Hmm, omitting `moduleVersion` results in the following compiler error: `org.apache.maven.plugins:maven-compiler-plugin:3.14.0:compile (compile-java9) on project truth: Fatal error compiling: error: bad value for --module-version option: 'HEAD-SNAPSHOT'`

Are you okay with changing the project version from `HEAD-SNAPSHOT` to `1.4.6-SNAPSHOT`?

cpovirk replied:
I have a possibly irrational dislike of the common style of `1.4.6-SNAPSHOT`:
- It "requires" that we know what the next version will be, but in fact we typically don't know whether it will be 1.4.6 or 1.5.0. (2.0.0, at least, is unlikely now that we've learned a lesson or two from Guava at the world's expense.)
- It means that anyone who is running a command on different snapshots (maybe because I want to rerun a command from a few years ago or maybe because I'm running `git bisect`) has to adjust the filename or do some awkward globbing instead of just hard-coding `HEAD-SNAPSHOT` or similar. (And things get even worse if that person doesn't use `clean` between builds.)

[In Guava](https://github.com/google/guava/blob/b94bd32f1923996261c0363b0a57283be391d87d/pom.xml#L9), we ended up with `999.0.0-SNAPSHOT`. That came after we encountered [the same problem as you've discovered](https://github.com/jspecify/jspecify/commit/0d39a0e8bf4ec863a2738835bb2421b25c48f282) with `HEAD-SNAPSHOT` (https://github.com/google/error-prone/pull/4311#issuecomment-1986230314). We'd considered `0.0.0-HEAD-SNAPSHOT` or `1.0.0-HEAD-SNAPSHOT` but anticipated/encountered trouble [with japicamp](https://github.com/google/error-prone/pull/4311#issuecomment-2256389047) and [with JDiff](https://github.com/google/guava/pull/7682), so we went the `999` route.

I'm not sure if I gave proper consideration to removing `HEAD` at that point. If `999.0.0-SNAPSHOT` sits better with you than `999.0.0-HEAD-SNAPSHOT`, then feel free to go for that.

cowwoc replied:
Well, yes and no. The `1.4.6-SNAPSHOT` style is asking you to take an educated guess of what the number version number will be. It doesn't matter if you're wrong, so long as you make sure to use the correct number before publishing to Maven Central. Until that time, all you're specifying is the version number that local builds should use... and for that purpose, any valid number should do.

I personally prefer `1.11.1-SNAPSHOT` to `999.0.0-HEAD-SNAPSHOT` because if someone mistakenly ends up publishing that high number to Maven Central, your artifact will be screwed. Anyone looking up your artifact will be shown that artifact as the latest release, even if lower-numbered releases are made at a later time.

Anyway, I'll move forward with `999.0.0-SNAPSHOT` for now :)

cpovirk commented on core/pom.xml:
```diff
@@ -98,6 +98,23 @@
+                <compileSourceRoot>${project.basedir}/src/main/java9</compileSourceRoot>
+              </compileSourceRoots>
+              <multiReleaseOutput>true</multiReleaseOutput>
+              <useIncrementalCompilation>false</useIncrementalCompilation>
```
I would have hoped that this would cause Maven to recompile everything every time, but it sounds like it actually causes it to [still attempt incremental compilation but in an even worse way](https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#useIncrementalCompilation)? I assume that the motivation is that it saves Maven from checking for other files that could be affected by a change to `module-info.java`, since we know that there should be no such files? I'd probably still leave it out, since it has this misleading name, but I could also see including it but with a comment to clarify its behavior. Is there more than I should be thinking about here?

cpovirk commented on core/src/main/java9/module-info.java:
```diff
@@ -0,0 +1,27 @@
+    requires static org.jspecify;
+    requires static com.google.errorprone.annotations;
+    requires static org.objectweb.asm;
+    requires static auto.value.annotations;
```
Hmm, we should probably at least get an Automatic-Module-Name into this artifact so that we know this name won't change over time (https://github.com/google/auto/issues/597). I'm a little fuzzy on the details of when an Automatic-Module-Name is good enough and when we need a `module-info`. Gemini tells me that a `module-info` is necessary for jlink but perhaps not for more "normal" use cases? But if we're going to do something to `auto-value-annotations`, maybe we should "just" modularize it, too?

cowwoc replied:
To my understanding `module-info.java` supersedes `Automatic-Module-Name` in every way. When possible, modularize all the way.

I'll try publishing a similar PR for https://github.com/google/auto/ shortly.

cpovirk commented on extensions/liteproto/pom.xml:
```diff
@@ -96,6 +96,31 @@
             </path>
           </annotationProcessorPaths>
         </configuration>
+        <executions>
```
It looks like the core Truth build is the only one with `useIncrementalCompilation` or `moduleVersion`. Once we're settled on what to use there, is it worth using for the others? (It's probably possible to share the configuration with `dependencyManagement`, but I don't really have an opinion either way on that.)

cpovirk reviewed (commented):
Thanks, I should have thought of this after the Guava modularization landed.

scordio commented:
If I may give a suggestion, I find [the setup demonstrated by Maven](https://maven.apache.org/plugins/maven-compiler-plugin/examples/module-info.html#java-6-to-8-compatibility) more flexible, i.e., Java 9 by default and one extra compilation with Java 8 excluding the module descriptor to make sure no Java 9 API is used.

Plus, it's IDE-friendly because the module descriptor is loaded properly.

See also a concrete example [here](https://github.com/scordio/jimfs-junit-jupiter/blob/b15ed0c495503a022b802aaea99405d8228465e8/pom.xml#L215-L238).

cowwoc commented:
@scordio Good point. I'll give that a try.

cowwoc committed:
Fix JPMS split package violations and enable documentation generation

Resolves module system conflicts by moving liteproto extension from shared
com.google.common.truth.extensions.proto package to unique
com.google.common.truth.extensions.liteproto namespace. This eliminates
split package violations that prevented proper module boundaries.

Enables aggregated Javadoc generation using legacyMode=true to work around
protobuf dependency conflicts between liteproto (protobuf-lite:3.0.1) and
proto (protobuf-java:4.32.1) extensions that caused module-path failures.

cowwoc committed:
Update copyright year to 2025 for newly created file

cowwoc commented:
It looks like I botched up the latest commit. Update coming soon.

cowwoc committed:
Make module-info compilation CI-safe

- Separate CI build (JDK 11) from test phases (JDK 8/11/17)
- Centralize module-info compilation config in parent POM
- Use two-stage compilation: module-info with release=9, base code with Java 8
- Maintain JDK 8 runtime compatibility via multi-release JAR structure

cowwoc committed:
Restore package documentation for proto extensions

- Add package-info.java for com.google.common.truth.extensions.proto
- Add package-info.java for com.google.common.truth.extensions.liteproto
- Clarify differences between full protobuf and protobuf-lite extensions
- Maintain proper Javadoc coverage for exported packages

These were inadvertently removed during JPMS package separation.

cowwoc commented:
Okay, I think I fixed all the problems. Please review.

cpovirk reviewed (commented):
As I suggest once or twice below, please do feel free to scope this down if that is sufficient to accomplish your goals.

cowwoc commented on core/src/main/java/module-info.java:
```diff
@@ -0,0 +1,29 @@
+    requires static org.jspecify;
+    requires static com.google.errorprone.annotations;
+    requires static org.objectweb.asm;
+    requires static auto.value.annotations;
```
Don't we need to wait for auto-value to get modularized before this will work?

cowwoc commented on core/src/main/java/module-info.java:
```diff
@@ -0,0 +1,29 @@
+
+module com.google.truth {
+    requires com.google.common;
+    requires junit;
```
Sounds like a good idea!

cowwoc committed:
Update core/src/main/java/module-info.java

Make the JUnit 4 dependency optional

Co-authored-by: Chris Povirk <cpovirk@google.com>

cpovirk commented on extensions/liteproto/src/main/java/com/google/common/truth/extensions/liteproto/LiteProtoSubject.java:
```diff
@@ -14,18 +14,19 @@
  */
 
-package com.google.common.truth.extensions.proto;
+package com.google.common.truth.extensions.liteproto;
```
Oh, of course, we can't modularize "for real" until Protobuf does :(

I would really like to avoid duplicating all the existing classes, even those just in the protobuf extensions. Truth (especially ProtoTruth) doesn't get extended the way that Protobuf does, but it would still mean two parallel mini-ecosystems. I'd like to think that we can work out the merging of the two protobuf artifacts if it comes to that.

cpovirk committed:
Fix module-info setup and GWT compatibility issues

cpovirk committed:
Merge remote-tracking branch 'origin/master' into add-module-info

# Conflicts:
#	.github/workflows/ci.yml

cpovirk committed:
Remove unnecessary java.compiler requirement from module-info.java

cpovirk committed:
test under 25 instead of 11, use 999.0.0-SNAPSHOT throughout instead of mixing with HEAD-SNAPSHOT, sort and format module-info

cpovirk committed:
Restore Employee sample extension and fix modular test visibility

cpovirk committed:
document add-exports

cpovirk committed:
remove autovalue annotations from module

cpovirk committed:
Move modular build configuration and GWT exclusions to core/pom.xml

cpovirk committed:
restore comment about why add-exports

cpovirk committed:
add transitive for junit, where it is probably important for users of Expect, and annotations, where it might not be but it seems possibly more correct and I would like to know of problems: https://github.com/jspecify/jspecify/issues/495#issuecomment-2744405987

cpovirk committed:
transitive for guava, too

cpovirk committed:
use java 26 for builds

cpovirk committed:
pin artifacts

cpovirk committed:
Fix CI failure by moving modular compilation to a JDK 9+ profile

cpovirk committed:
Skip main compilation in CI test job to preserve build artifacts

cpovirk committed:
Unpack build artifacts in CI test job to satisfy test-compilation

cpovirk committed:
Merge branch 'master' into add-module-info
